File: Sorting_KlastaKwik/klustakwik_Linux/batchKwik.m
%%
% script for batch running KlustaKwik in linux
% 20240713 tjf

%% all folders
folders = { ...
    '/media/lab008/Jeffrey_3/LE26_240705'; ...
    '/media/lab008/Jeffrey_3/LE26_240706'; ...
    '/media/lab008/Jeffrey_3/LE26_240707'  ...
    };
interval = 0;  % interval before processing next folder

%%
cwd = pwd;
nFolder = size(folders, 1);
for iFolder = 1: nFolder
    folder = folders{iFolder};
    cd(folder);
    fetFile = dir('*.fet.*');
    loc = strfind(fetFile(1).name, '.');
    if ~iscell(fetFile)
        for iFile = 1: numel(fetFile)
            c_idle = check_load_linux;
            while c_idle < 10
                disp(['System resource is low now(' num2str(c_idle) '%), will wait for 10 seconds to resubmit the job.']);
                pause(10)
                c_idle=check_load_linux;
            end
            system(['KlustaKwik ' fetFile(1).name(1:loc(end-1)-1) ' ' num2str(iFile) ...
                ' -MinClusters 20-MaxClusters 30 '...
                ' -MaxPossibleClusters 50&' ]);
            disp(['Job submitted ' num2str(iFile) '/' num2str(length(fetFile)) ' current system resource:' num2str(c_idle) '%']);
        end
    end
    pause(interval);
end
cd(cwd);

File: Sorting_KlastaKwik/klustakwik_Windows/batch_kwik.py
import os
import time
import subprocess
import re

# 所有路径
f_kwik = r"C:\path\to\KlustaKwik"
folders = [
    r"C:\path\to\data\LE16_240124",
    r"C:\path\to\data\LE16_240125"
]
interval = 0  # 处理下一个文件夹前的间隔时间

def check_load_windows():
    try:
        result = subprocess.run(["wmic", "cpu", "get", "loadpercentage"], capture_output=True, text=True)
        lines = result.stdout.splitlines()
        cpu_loads = [int(s) for s in lines if s.strip().isdigit()]
        cpu_idle = 100 - cpu_loads[0] if cpu_loads else 100
    except Exception:
        cpu_idle = 100
    return cpu_idle

cwd = os.getcwd()
for folder in folders:
    os.chdir(folder)
    fet_files = [f for f in os.listdir(folder) if re.match(r'.+\.fet\.\d+$', f)]
    
    for i, fet_file in enumerate(fet_files, start=1):
        c_idle = check_load_windows()
        while c_idle < 10:
            print(f'System resource is low now ({c_idle}%), will wait for 10 seconds to resubmit the job.')
            time.sleep(10)
            c_idle = check_load_windows()
        
        base_name = fet_file.rsplit('.', 2)[0]  # 截取到 .fet 之前的部分
        command = f'{f_kwik} {base_name} {i} -MinClusters 20 -MaxClusters 30 -MaxPossibleClusters 50 &'
        subprocess.run(command, shell=True)
        print(f'Job submitted {i}/{len(fet_files)} current system resource: {c_idle}%')
    
    time.sleep(interval)

os.chdir(cwd)
